Phicomm M1 air sensor

Yonsm opened the pull request:
## Description:


**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5142

## Example entry for `configuration.yaml` (if applicable):
```yaml
sensors:
  - platform: phicomm
    name: Phicomm
    username: ********
    password: ********
    sensors: [pm25, hcho, temperature, humidity]
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

Commit:
Add Phicomm air sensor

Commit:
Refine with script/lint

Commit:
Exclude broadlink.py in .coveragerc

Kane610 commented on homeassistant/components/sensor/phicomm.py:
```diff
@@ -0,0 +1,196 @@
+        return None
+
+
+class PhicommData():
```
This should be moved to an external package and hosted on pypi.
With little effort this could also be made async,

Commit:
Update with async mode

1.  Async mode to setup platform.
2. Async update by async_track_time_interval;
3. Discard devices conf, detect device count from server data;.

Commit:
Resolve lint error

Yonsm commented:
@Kane610 Update to async mode. but have not been moved to pypi. The data worker is too simple, maybe is it optional to move to pypi?

Kane610 commented:
It is a requirement from hass that you host API specific code on pypi see https://www.home-assistant.io/developers/code_review_platform/#6-communication-with-devicesservices

Kane610 commented on homeassistant/components/sensor/phicomm.py:
```diff
@@ -0,0 +1,228 @@
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
```
HASS requires at least python 3.5.3, so it is safe to use the async/await syntax

Kane610 commented on homeassistant/components/sensor/phicomm.py:
```diff
@@ -0,0 +1,228 @@
+    username = config.get(CONF_USERNAME)
+    password = config.get(CONF_PASSWORD)
+    sensors = config.get(CONF_SENSORS)
+    scan_interval = config.get(CONF_SCAN_INTERVAL)
```
All of these parameters can be expected to be available as they are either required or you have provided a default value so there is no need to do get

Kane610 commented on homeassistant/components/sensor/phicomm.py:
```diff
@@ -0,0 +1,228 @@
+    phicomm = PhicommData(hass, username, password)
+    devices = yield from phicomm.make_sensors(name, sensors)
+    if devices:
+        async_add_devices(devices)
```
If this is async you should be await/yield from right?

Kane610 commented on homeassistant/components/sensor/phicomm.py:
```diff
@@ -0,0 +1,228 @@
+    def make_sensors(self, name, sensors):
+        """Make sensors with online data."""
+        try:
+            with open(self._token_path) as file:
```
you need to await/yield from any IO operation

Kane610 replied:
This is still an I/O operation, which needs to be async

Kane610 commented on homeassistant/components/sensor/phicomm.py:
```diff
@@ -0,0 +1,228 @@
+            else:
+                return None
+        headers = {'User-Agent': USER_AGENT, 'Authorization': self._token}
+        return requests.get(DATA_URL, headers=headers).json()
```
requests isn't async, you need to look into aiohttp, you should also use hass' aio session and not create your own

Kane610 commented:
Right now you're selectively async, so you're halfway there! You need to change from requests to aiohttp and preferably use the new async/await syntax.

Commit:
Async/await syntax

Kane610 commented on homeassistant/components/sensor/phicomm.py:
```diff
@@ -57,22 +55,29 @@
+    sensors = config[CONF_SENSORS]
+    scan_interval = config[CONF_SCAN_INTERVAL]
 
     phicomm = PhicommData(hass, username, password)
```
Since this should be broken out of hass you should also have it decoupled from hass; you only need to pass the loop, session object and token path

Kane610 commented on homeassistant/components/sensor/phicomm.py:
```diff
@@ -150,79 +155,67 @@ def __init__(self, hass, username, password):
+                    'password': md5.hexdigest().upper()}
             headers = {'User-Agent': USER_AGENT}
-            json = requests.post(TOKEN_URL, headers=headers, data=data).json()
+            async with session.post(TOKEN_URL, headers=headers, data=data) \
```
error handling

Kane610 commented on homeassistant/components/sensor/phicomm.py:
```diff
@@ -150,79 +155,67 @@ def __init__(self, hass, username, password):
+
         headers = {'User-Agent': USER_AGENT, 'Authorization': self._token}
-        return requests.get(DATA_URL, headers=headers).json()
+        async with session.get(DATA_URL, headers=headers) as response:
```
error handling

Yonsm replied:
I am not familar with Python. Just refer some other source code from ha release. They do the same code, so did I. Sorry for my non-professional code. If you can figure out the correct
 way, your help will be highly appreciated.

Kane610 replied:
You're already using try and exception handling in some parts, you just need it for the rest as well

Kane610 commented on homeassistant/components/sensor/phicomm.py:
```diff
@@ -150,79 +155,67 @@ def __init__(self, hass, username, password):
                 return None
+
+            self._token = json['access_token']
+            with open(self._token_path, 'w') as file:
```
should be async

Commit:
Async token reading & aiohttp error handling

MartinHjelmare commented on homeassistant/components/sensor/phicomm.py:
```diff
@@ -0,0 +1,232 @@
+            import traceback
+            _LOGGER.error('Exception: %s', traceback.format_exc())
+
+    async def fetch_data(self):
```
Data fetching and parsing should be moved to standalone library and published on pypi.

fabaff commented:
This PR seems to have gone stale. Closing it. You can reopen it when you're ready to finish it.
